src/app/nav/events.cljs

```diff
@@ -1,21 +1,49 @@
 (ns app.nav.events
-  (:require [re-frame.core :refer [reg-event-db reg-fx]]
+  (:require [re-frame.core :refer [reg-event-db reg-event-fx reg-fx path]]
             [app.router :as router]))
 
+(def nav-interceptors [(path :nav)])
+
 (reg-fx
  :navigate-to
  (fn [{:keys [path]}]
    (router/set-token! path)))
 
-(reg-event-db
-  :route-changed
-  (fn [db [_ {:keys [handler route-params]}]]
-    (-> db
-        (assoc-in [:nav :active-page] handler)
-        (assoc-in [:nav :active-income] (keyword (:income-id route-params)))
-        (assoc-in [:nav :active-vendor] (keyword (:vendor-id route-params)))
-        (assoc-in [:nav :active-category] (keyword (:category-id route-params)))
-        (assoc-in [:nav :active-bill] (keyword (:bill-id route-params))))))
+(reg-event-fx
+ :route-changed
+ nav-interceptors
+ (fn [{nav :db} [_ {:keys [handler route-params]}]]
+   (let [nav (assoc nav :active-page handler)]
+     (case handler
+       :incomes
+       {:db nav
+        :dispatch [:fetch-incomes]}
+
+       :income
+       {:db (assoc nav :active-income (keyword (:income-id route-params)))}
+
+       :vendors
+       {:db nav
+        :dispatch [:fetch-vendors]}
+
+       :vendor
+       {:db (assoc nav :active-vendor (keyword (:vendor-id route-params)))}
+
+       :bill
+       {:db (assoc nav :active-bill (keyword (:bill-id route-params)))}
+
+       :categories
+       {:db nav
+        :dispatch [:fetch-categories]}
+
+       :category
+       {:db (assoc nav :active-category (keyword (:category-id route-params)))}
+
+       {:db (dissoc nav
+                    :active-income
+                    :active-vendor
+                    :active-category
+                    :active-bill)}))))
```

Comment:
i think it is fine, maybe big and contains a lot of 'logic' but it is simple, easy to debug and easy to reason about, until this grows significantly or you run into some troubles i would not complicate it anymore

if it grows too big, it would probably be good to convert this function to multimethod that will dispatch on handler

also, not sure about `:income` for example - does it just use the previously loaded income or the fetch logic is somewehre inside the income page? if the former, will this still work if you just paste a concrete income url? because there will be no incomes in db yet?